File: mem0/utils/scoring.py
"""
Scoring utilities for hybrid retrieval.

Provides:
- **BM25 normalization**: Sigmoid normalization of raw BM25 scores to [0, 1].
- **BM25 parameter selection**: Query-length-adaptive sigmoid parameters.
- **Additive scoring**: Combined scoring with semantic + BM25 + entity boost.
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Optional


def get_bm25_params(query: str, *, lemmatized: Optional[str] = None) -> tuple:
    """Get BM25 sigmoid parameters based on query length.

    Longer queries tend to have higher raw BM25 scores, so we adjust
    the sigmoid midpoint and steepness accordingly.

    Returns:
        (midpoint, steepness) for sigmoid normalization.
    """
    if lemmatized is None:
        from mem0.utils.lemmatization import lemmatize_for_bm25

        lemmatized = lemmatize_for_bm25(query)
    num_terms = len(lemmatized.split()) if lemmatized else 1

    if num_terms <= 3:
        return 5.0, 0.7
    elif num_terms <= 6:
        return 7.0, 0.6
    elif num_terms <= 9:
        return 9.0, 0.5
    elif num_terms <= 15:
        return 10.0, 0.5
    else:
        return 12.0, 0.5


def normalize_bm25(raw_score: float, midpoint: float, steepness: float) -> float:
    """Normalize BM25 score to [0, 1] using logistic sigmoid.

    Args:
        raw_score: Raw BM25 score (unbounded, typically 0-20+).
        midpoint: Score at which sigmoid outputs 0.5.
        steepness: Controls how quickly sigmoid transitions.

    Returns:
        Normalized score in range [0, 1].
    """
    return 1.0 / (1.0 + math.exp(-steepness * (raw_score - midpoint)))


ENTITY_BOOST_WEIGHT = 0.5


def score_and_rank(
    candidates: List[Dict[str, Any]],
    bm25_scores: Dict[str, float],
    entity_boosts: Dict[str, float],
    threshold: float,
    top_k: int,
    explain: bool = False,
) -> List[Dict[str, Any]]:
    """Score candidates additively and return top-k results.

    The candidate pool is the union of semantic, keyword, and entity-linked
    memories.  A candidate that was only found via keyword or entity boost
    (no semantic hit) has semantic_score = 0.0.

    Threshold gating:
        - Candidates with a semantic score pass if semantic >= threshold.
        - Candidates *without* a meaningful semantic score (pure keyword or
          pure entity) pass if they have at least one non-semantic signal
          (bm25 > 0 or entity_boost > 0).  This prevents purely semantic
          low-quality hits from leaking through while allowing keyword /
          entity-only matches to participate in ranking.

    Combined score:
        combined = (semantic + bm25 + entity_boost) / max_possible

    The divisor adapts based on which signals are active:
        - Semantic only: max_possible = 1.0
        - Semantic + BM25: max_possible = 2.0
        - Semantic + BM25 + entity: max_possible = 2.5
        - Semantic + entity (no BM25): max_possible = 1.5
        - BM25 only (no semantic): max_possible = 1.0
        - Entity only (no semantic): max_possible = ENTITY_BOOST_WEIGHT

    Args:
        candidates: Unified candidate pool (semantic + keyword + entity).
            Each dict must have "id", and may have "score" (semantic),
            "payload", "bm25_score", and "entity_boost".
        bm25_scores: Normalized keyword scores keyed by memory ID.
        entity_boosts: Entity-link boosts keyed by memory ID.
        threshold: Minimum semantic score required for semantic-only candidates.
        top_k: Maximum number of results to return.
        explain: Include score_details in each result when true.

    Returns:
        List of scored result dicts sorted by combined score descending.
    """
    has_bm25 = bool(bm25_scores)
    has_entity = bool(entity_boosts)

    scored: List[Dict[str, Any]] = []

    for result in candidates:
        mem_id = result.get("id")
        if mem_id is None:
            continue

        mem_id_str = str(mem_id)
        semantic_score = result.get("score") or 0.0
        bm25_score = bm25_scores.get(mem_id_str, result.get("bm25_score", 0.0))
        entity_boost = entity_boosts.get(mem_id_str, result.get("entity_boost", 0.0))

        has_semantic = semantic_score > 0.0
        has_non_semantic = bm25_score > 0.0 or entity_boost > 0.0

        if has_semantic and semantic_score < threshold and not has_non_semantic:
            continue

        if not has_semantic and not has_non_semantic:
            continue

        active_max = 0.0
        if has_semantic:
            active_max += 1.0
        if has_bm25 and bm25_score > 0.0:
            active_max += 1.0
        if has_entity and entity_boost > 0.0:
            active_max += ENTITY_BOOST_WEIGHT

        if active_max == 0.0:
            continue

        raw_combined = semantic_score + bm25_score + entity_boost
        combined = min(raw_combined / active_max, 1.0)

        scored_result = {
            "id": mem_id_str,
            "score": combined,
            "payload": result.get("payload"),
        }
        if explain:
            scored_result["score_details"] = {
                "semantic_score": semantic_score,
                "bm25_score": bm25_score,
                "entity_boost": entity_boost,
                "raw_score": raw_combined,
                "max_possible_score": active_max,
                "final_score": combined,
                "threshold": threshold,
            }
        scored.append(scored_result)

    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:top_k]

File: mem0-ts/src/oss/src/utils/scoring.ts
/**
 * Scoring utilities for hybrid retrieval.
 *
 * Provides:
 * - BM25 normalization: Sigmoid normalization of raw BM25 scores to [0, 1].
 * - BM25 parameter selection: Query-length-adaptive sigmoid parameters.
 * - Additive scoring: Combined scoring with semantic + BM25 + entity boost.
 */

export const ENTITY_BOOST_WEIGHT = 0.5;

export interface ScoreWeights {
  semanticWeight?: number;
  bm25Weight?: number;
  entityBoostWeight?: number;
}

export interface ScoreDetails {
  semanticScore: number;
  bm25Score: number;
  entityBoost: number;
  rawScore: number;
  maxPossibleScore: number;
  finalScore: number;
  threshold: number;
  weights?: Required<ScoreWeights>;
}

/**
 * Get BM25 sigmoid parameters based on query length.
 *
 * Longer queries tend to have higher raw BM25 scores, so we adjust
 * the sigmoid midpoint and steepness accordingly.
 *
 * @param query - The original query string.
 * @param lemmatized - Optional pre-lemmatized query string. If not provided,
 *   the term count is estimated from the raw query.
 * @returns A tuple of [midpoint, steepness] for sigmoid normalization.
 */
export function getBm25Params(
  query: string,
  lemmatized?: string,
): [number, number] {
  const text = lemmatized ?? query;
  const numTerms = text.trim().split(/\s+/).filter(Boolean).length || 1;

  if (numTerms <= 3) {
    return [5.0, 0.7];
  } else if (numTerms <= 6) {
    return [7.0, 0.6];
  } else if (numTerms <= 9) {
    return [9.0, 0.5];
  } else if (numTerms <= 15) {
    return [10.0, 0.5];
  } else {
    return [12.0, 0.5];
  }
}

/**
 * Normalize a raw BM25 score to [0, 1] using logistic sigmoid.
 *
 * @param rawScore - Raw BM25 score (unbounded, typically 0-20+).
 * @param midpoint - Score at which sigmoid outputs 0.5.
 * @param steepness - Controls how quickly sigmoid transitions.
 * @returns Normalized score in range [0, 1].
 */
export function normalizeBm25(
  rawScore: number,
  midpoint: number,
  steepness: number,
): number {
  return 1.0 / (1.0 + Math.exp(-steepness * (rawScore - midpoint)));
}

export interface ScoredResult {
  id: string;
  score: number;
  payload: Record<string, any>;
  scoreDetails?: ScoreDetails;
}

/**
 * Score candidates additively and return top-k results.
 *
 * For each candidate:
 *   combined = (semantic * w_sem + bm25 * w_bm25 + entity_boost * w_ent) / max_possible
 *
 * Threshold gates the semantic score BEFORE combining -- candidates
 * below the threshold are excluded even if BM25/entity would boost them.
 *
 * The divisor adapts based on which signals are active and their weights.
 *
 * @param semanticResults - Candidate results with id, score, and payload.
 * @param bm25Scores - Map of memory ID to normalized BM25 score.
 * @param entityBoosts - Map of memory ID to entity boost score.
 * @param threshold - Minimum semantic score to include a candidate.
 * @param topK - Maximum number of results to return.
 * @param explain - Include scoreDetails in each result when true.
 * @param weights - Custom weights for each score component.
 * @returns Sorted list of scored results, highest score first.
 */
export function scoreAndRank(
  semanticResults: Array<{
    id: string;
    score: number;
    payload: Record<string, any>;
  }>,
  bm25Scores: Record<string, number>,
  entityBoosts: Record<string, number>,
  threshold: number,
  topK: number,
  explain: boolean = false,
  weights: ScoreWeights = {},
): ScoredResult[] {
  const hasBm25 = Object.keys(bm25Scores).length > 0;
  const hasEntity = Object.keys(entityBoosts).length > 0;

  const hasCustomWeights =
    weights.semanticWeight !== undefined ||
    weights.bm25Weight !== undefined ||
    weights.entityBoostWeight !== undefined;

  const semanticWeight = weights.semanticWeight ?? 1.0;
  const bm25Weight = weights.bm25Weight ?? 1.0;
  const entityBoostWeight = hasCustomWeights
    ? (weights.entityBoostWeight ?? ENTITY_BOOST_WEIGHT)
    : 1.0;

  const maxEntityComponent = hasCustomWeights
    ? entityBoostWeight
    : ENTITY_BOOST_WEIGHT;

  let maxPossible = semanticWeight;
  if (hasBm25) {
    maxPossible += bm25Weight;
  }
  if (hasEntity) {
    maxPossible += maxEntityComponent;
  }

  const appliedWeights: Required<ScoreWeights> = {
    semanticWeight,
    bm25Weight,
    entityBoostWeight: hasCustomWeights
      ? entityBoostWeight
      : ENTITY_BOOST_WEIGHT,
  };

  const scored: ScoredResult[] = [];

  for (const result of semanticResults) {
    const memId = result.id;
    if (memId == null) {
      continue;
    }

    const semanticScore = result.score ?? 0.0;
    if (semanticScore < threshold) {
      continue;
    }

    const memIdStr = String(memId);
    const bm25Score = bm25Scores[memIdStr] ?? 0.0;
    const entityBoost = entityBoosts[memIdStr] ?? 0.0;

    const weightedSemantic = semanticScore * semanticWeight;
    const weightedBm25 = bm25Score * bm25Weight;
    const weightedEntity = entityBoost * entityBoostWeight;

    const rawCombined = weightedSemantic + weightedBm25 + weightedEntity;
    const combined = Math.min(rawCombined / maxPossible, 1.0);

    const entry: ScoredResult = {
      id: memIdStr,
      score: combined,
      payload: result.payload,
    };
    if (explain) {
      entry.scoreDetails = {
        semanticScore,
        bm25Score,
        entityBoost,
        rawScore: rawCombined,
        maxPossibleScore: maxPossible,
        finalScore: combined,
        threshold,
        weights: appliedWeights,
      };
    }
    scored.push(entry);
  }

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK);
}
